README.md
# RandomLSystem generator
haskell practise, random LSystem generator and displayer


ChangeLog.md
# Changelog for RandomLSystem

## Unreleased changes


src/LSystem.hs
module LSystem (
    LSystem (LSystem),
    State, Rules,
    stateAfterNSteps, translateState, translateAfterNSteps
               ) where

import qualified Data.Map.Strict as Map
import Data.Maybe

type State a = [a]
type Rules a b = [(a, [b])]
type RulesMap a b = Map.Map a [b]

data LSystem a b = LSystem {
    variables            :: [a],
    constants            :: [a],
    axiom                :: State a,
    rules                :: Rules a a,
    translationRules     :: Rules a b
                           } deriving (Show, Read)

doOneStep :: Ord a => RulesMap a a -> State a -> State a
doOneStep rulesMap = concatMap (fromMaybe [] . (rulesMap Map.!?))

stateAfterNSteps :: Ord a => LSystem a b -> Int -> State a
stateAfterNSteps (LSystem vars cons axiom rules tr) n = iterate (doOneStep fullRules) axiom !! n
    where
        fullRules = foldr (\f m -> f m) (Map.fromList rules) [Map.insert x [x] | x <- cons]

translateState :: Eq a => LSystem a b -> State a -> [b]
translateState (LSystem v c a r tr) state = concatMap fromJust $ filter isJust $ map (`lookup` tr) state

translateAfterNSteps :: Ord a => LSystem a b -> Int -> [b]
translateAfterNSteps lsys@(LSystem v c a r tr) n = concatMap fromJust $ filter isJust $ map (`lookup` tr) (stateAfterNSteps lsys n)


src/TurtleDraw.hs
module TurtleDraw (
    display,
    Command (..),
    Pen (..), black, white, red, green, blue, yellow,
    Distance, Angle,
    triangle, tree,
    join
   )
    where

import qualified Graphics.Rendering.OpenGL as GL
import Graphics.UI.GLUT hiding (Angle)
import Data.IORef
import Data.List
import Control.Monad( liftM, liftM2, liftM3 )
import Test.QuickCheck

infixr 5 :#:


join :: [Command] -> Command
join [] = Sit
join list = foldr1 (:#:) list


-- Points

data Pnt = Pnt Float Float
  deriving (Eq,Ord,Show)

instance Num Pnt where
  Pnt x y + Pnt x' y'  =  Pnt (x+x') (y+y')
  Pnt x y - Pnt x' y'  =  Pnt (x-x') (y-y')
  Pnt x y * Pnt x' y'  =  Pnt (x*x') (y*y')
  fromInteger          =  scalar . fromInteger
  abs (Pnt x y)        =  Pnt (abs x) (abs y)
  signum (Pnt x y)     =  Pnt (signum x) (signum y)

instance Fractional Pnt where
  Pnt x y / Pnt x' y'  =  Pnt (x/x') (y/y')
  fromRational         =  scalar . fromRational

scalar :: Float -> Pnt
scalar x  =  Pnt x x

scalarMin :: Pnt -> Pnt
scalarMin (Pnt x y)  =  scalar (x `min` y)

scalarMax :: Pnt -> Pnt
scalarMax (Pnt x y)  =  scalar (x `max` y)

dimensions :: Pnt -> (Int,Int)
dimensions (Pnt x y)  =  (ceiling x, ceiling y)

lub :: Pnt -> Pnt -> Pnt
Pnt x y `lub` Pnt x' y'  =  Pnt (x `max` x') (y `max` y')

glb :: Pnt -> Pnt -> Pnt
Pnt x y `glb` Pnt x' y'  =  Pnt (x `min` x') (y `min` y')

-- The last two functions are not called min and max
-- because the invariant for min and max states
-- (min x y, max x y) = (x,y) or (y,x).

pointToSize :: Pnt -> Size
pointToSize (Pnt x y) = Size (ceiling x) (ceiling y)

sizeToPoint :: Size -> Pnt
sizeToPoint (Size x y) = Pnt (fromIntegral x) (fromIntegral y)

-- Colors

data Pen = Colour GL.GLfloat GL.GLfloat GL.GLfloat
         | Inkless
           deriving (Eq, Ord, Show)

penToRGB :: Pen -> GL.Color3 GL.GLfloat
penToRGB (Colour r g b)  =  GL.Color3 r g b
penToRGB Inkless  =  error "penToRGB: inkless"

white, black, red, green, blue :: Pen
white = Colour 1.0 1.0 1.0
black = Colour 0.0 0.0 0.0
red   = Colour 1.0 0.0 0.0
green = Colour 0.0 1.0 0.0
blue  = Colour 0.0 0.0 1.0
yellow  = Colour 1.0 1.0 0.0

-- Lines

data Ln = Ln Pen Pnt Pnt
  deriving (Eq,Ord,Show)

-- Window parameters

theCanvas :: Pnt
theCanvas  =  Pnt 800 800

theBGcolor :: GL.Color3 GL.GLfloat
theBGcolor = penToRGB white



-- Main drawing and window functions

display :: Command -> IO ()
display c = do
  initialDisplayMode $= [DoubleBuffered]
  initialWindowSize  $= pointToSize theCanvas
  getArgsAndInitialize
  w <- createWindow "Turtle Graphics"
  displayCallback $= draw c
  reshapeCallback $= Just (\x -> (viewport $= (Position 0 0, x)))
  --actionOnWindowClose $= ContinueExectuion
  draw c
  mainLoop

draw :: Command -> IO ()
draw c = do clear [ColorBuffer]
            loadIdentity
            background 
            toGraphic $ rescale $ execute c
            swapBuffers

toGraphic :: [Ln] -> IO ()
toGraphic lines  = sequence_ (map f lines)
  where
  f (Ln pen startP endP)  =
    GL.color (penToRGB pen) >>
    GL.renderPrimitive GL.LineStrip (toVertex startP >> toVertex endP)

background :: IO ()
background = do GL.color theBGcolor
                GL.renderPrimitive GL.Polygon $ mapM_ GL.vertex
                      [GL.Vertex3 (-1) (-1) 0,
                       GL.Vertex3   1  (-1) 0,
                       GL.Vertex3   1    1  0,
                       GL.Vertex3 (-1)   1 (0::GL.GLfloat) ]


toVertex (Pnt x y)  =  GL.vertex $ GL.Vertex3 
 (realToFrac x) (realToFrac y) (0::GL.GLfloat)



-- Commands for moving the turtle around
--  Turtles turn counter-clockwise and start facing up

type Angle    = Float
type Distance = Float
type Turtle   = (Pen,Angle,Pnt)

data Command = Go Distance
             | Turn Angle 
             | Sit
             | Command :#: Command
             | Branch Command
             | GrabPen Pen
               deriving (Eq, Ord, Show)


-- Converting commands to GL graphics

execute :: Command -> [Ln]
execute c  =  lines
  where
  (lines, turtle)  =  f c (black, 0, Pnt 0 0)

  f :: Command -> Turtle -> ([Ln], Turtle)
  f (c :#: d) turtle             =  (clines ++ dlines, dturtle)
                                    where
                                    (clines, cturtle) = f c turtle
                                    (dlines, dturtle) = f d cturtle
  f (Branch c) turtle            =  (clines, turtle)
                                    where
                                    (clines, cturtle) = f c turtle
  f (Go dst) (pen,ang,pnt)       =  (if pen == Inkless
                                       then []
                                       else [Ln pen pnt endpnt],
                                     (pen,ang,endpnt))
                                    where
                                    endpnt = pnt + scalar dst * polar ang
  f (Turn delta) (pen,ang,pnt)   =  ([], (pen,ang-delta,pnt))
  f (GrabPen new) (old,ang,pnt)  =  ([], (new,ang,pnt))
  f (Sit) turtle                 =  ([], turtle)


-- Rescales all points in a list of lines
--  from an arbitrary scale
--  to (-1.-1) - (1.1)

rescale :: [Ln] -> [Ln]
rescale lines | points == [] = []
              | otherwise    = map f lines
  where
  f (Ln pen p q)  =  Ln pen (g p) (g q)
  g p             =  swap ((p - p0) / s)
  points          =  [ r | Ln pen p q <- lines, r <- [p, q] ]
  hi              =  foldr1 lub points
  lo              =  foldr1 glb points
  s               =  scalarMax (hi - lo) * scalar (0.55)
  p0              =  (hi + lo) * scalar (0.5)
  swap (Pnt x y)  =  Pnt y x

polar :: Angle -> Pnt
polar ang  =  Pnt (cos radians) (sin radians)
  where
  radians  =  ang * 2 * pi / 360


-- Sample LSystems

triangle :: Int -> Command
triangle x  =  p :#: f x
  where
  f 0      = Go 10
  f x  = f (x-1) :#: p :#: f (x-1) :#: n :#: f (x-1) :#: n :#: f (x-1) :#: p :#: f (x-1)
  n        = Turn 90
  p        = Turn (-90)

tree :: Int -> Command
tree x  =  f x
  where
  f 0      = GrabPen red :#: Go 10
  f x  = g (x-1) :#: Branch (n :#: f (x-1))
                 :#: Branch (p :#: f (x-1))
                 :#: Branch (g (x-1) :#: f (x-1))
  g 0      = GrabPen blue :#: Go 10
  g x  = g (x-1) :#: g (x-1)
  n        = Turn 45
  p        = Turn (-45)



-- Generators for QuickCheck

instance Arbitrary Pen where
    arbitrary  =  sized pen
        where
          pen n  =  elements [black,red,green,blue,white,Inkless]


instance Arbitrary Command where
    arbitrary  =  sized cmd 
        where
          cmd n  |  n <= 0     =  oneof [liftM (Go . abs) arbitrary,
                                         liftM Turn arbitrary ]
                 |  otherwise  =  liftM2 (:#:) (cmd (n `div` 2)) (cmd (n `div`2))





src/TurtleDrawGLUT.hs
module TurtleDrawGLUT (
    Command (..),
    white, black, red, green, blue,
    showLSystem
                      ) where

import Graphics.UI.GLUT hiding (Line)
import Data.IORef
import Data.StateVar

import LSystem


data Command = Go Float | Turn Float | SetColor (Float, Float, Float) | Branch [Command]
    deriving (Show, Read)

--                       position    angle     color
data Turtle = Turtle (Vertex2 Float) Float (Color3 Float)
    deriving (Show)

data Line = Line (Color3 Float) (Vertex2 Float) (Vertex2 Float)
    deriving (Show)

-- Lets predefine some colors
white, black, red, green, blue :: (Float, Float, Float)
white = (255, 255, 255)
black = (0  , 0  , 0  )
red   = (255, 0  , 0  )
green = (0  , 255, 0  )
blue  = (0  , 0  , 255)

commandsToLines :: [Command] -> [Line]
commandsToLines = turtlePathLines (Turtle (Vertex2 0 0) 0 (Color3 255 255 255))
    where
        turtlePathLines :: Turtle -> [Command] -> [Line]
        turtlePathLines t [] = []
        turtlePathLines (Turtle tp@(Vertex2 tx ty) ta tc) (Go d:cs) = Line tc tp newTp : turtlePathLines (Turtle newTp ta tc) cs
            where
                aRad = (pi / 180) * ta
                newTp = Vertex2 (tx + d*sin aRad) (ty + d*cos aRad)
        turtlePathLines (Turtle tp ta tc) (Turn a:cs) = turtlePathLines (Turtle tp (ta + a) tc) cs
        turtlePathLines (Turtle tp ta tc) (SetColor (r, g, b):cs) = turtlePathLines (Turtle tp ta (Color3 r g b)) cs
        turtlePathLines t (Branch bcs:cs) = turtlePathLines t bcs ++ turtlePathLines t cs

scaleLine :: Float -> Line -> Line
scaleLine k (Line c v1 v2) = Line c (scaleV k v1) (scaleV k v2)
    where
        scaleV k (Vertex2 vx vy) = Vertex2 (vx * k) (vy * k)

rescaleLines :: [Line] -> [Line]
rescaleLines lines = map (scaleLine k) lines
    where
        maxD = foldr (\(Line c (Vertex2 v1x v1y) (Vertex2 v2x v2y)) p -> maximum $ map abs [p, v1x, v1y, v2x, v2y]) 0 lines
        k = 0.9 * (1 / maxD)

renderLine :: Line -> IO ()
renderLine (Line col ver1 ver2)= renderPrimitive Lines $ do
    color col
    vertex ver1
    vertex ver2

showLSystem :: Ord a => LSystem a Command-> IO ()
showLSystem lsys = do
    initialWindowSize $= Size 800 800
    (_progName, _args) <- getArgsAndInitialize
    _window <- createWindow "Showing a LSystem"

    nIORef <- newIORef (0 :: Int)
    nIORef $= 5

    displayCallback $= showLSystemDisplay lsys nIORef
    keyboardCallback $= Just (showLSystemKeyboard nIORef)
    mouseWheelCallback $= Just (showLSystemMouseWheel nIORef)
    actionOnWindowClose $= ContinueExecution

    mainLoop

showLSystemKeyboard :: IORef Int -> KeyboardCallback
showLSystemKeyboard nIORef c _ = case c of
    '-' -> nIORef $~ max 0 . (+(-1)) >> postRedisplay Nothing
    '+' -> nIORef $~ min 15 . (+1) >> postRedisplay Nothing

showLSystemMouseWheel :: IORef Int -> MouseWheelCallback
showLSystemMouseWheel nIORef _ d _ = do
    nIORef $~ max 1 . (+d)

showLSystemDisplay :: Ord a => LSystem a Command -> IORef Int -> DisplayCallback
showLSystemDisplay lsys nIORef = do
    clear [ColorBuffer]
    n <- get nIORef
    mapM_ renderLine $ rescaleLines $ commandsToLines $ translateAfterNSteps lsys n
    renderString TimesRoman24 (show n)
    flush


src/LSystemRenderer.hs
module LSystemRenderer (
    Command (..),
    white, black, red, green, blue,
    showCommands, showLSystem
                      ) where

import Graphics.UI.GLUT hiding (Line)
import Data.IORef
import Data.StateVar
import Data.Maybe

import LSystem


data Command = Go Float | Turn Float | SetColor (Float, Float, Float) | StartBranch | EndBranch
    deriving (Show, Read)

--                       position    angle     color
data Turtle = Turtle (Vertex2 Float) Float (Color3 Float)
    deriving (Show)

data Line = Line (Color3 Float) (Vertex2 Float) (Vertex2 Float)
    deriving (Show)

-- Lets predefine some colors
white, black, red, green, blue :: (Float, Float, Float)
white = (255, 255, 255)
black = (0  , 0  , 0  )
red   = (255, 0  , 0  )
green = (0  , 255, 0  )
blue  = (0  , 0  , 255)

commandsToLines :: [Command] -> [Line]
commandsToLines = fst . turtlePathLines (Turtle (Vertex2 0 0) 0 (Color3 255 255 255))
    where
        turtlePathLines :: Turtle -> [Command] -> ([Line], Maybe [Command])
        turtlePathLines t [] = ([], Nothing)
        turtlePathLines (Turtle tp@(Vertex2 tx ty) ta tc) (Go d:cs) = (Line tc tp newTp : linesAfter, maybeBranchCs)
            where
                aRad = (pi / 180) * ta
                newTp = Vertex2 (tx + d*sin aRad) (ty + d*cos aRad)
                (linesAfter, maybeBranchCs) = turtlePathLines (Turtle newTp ta tc) cs
        turtlePathLines (Turtle tp ta tc) (Turn a:cs) = turtlePathLines (Turtle tp (ta + a) tc) cs
        turtlePathLines (Turtle tp ta tc) (SetColor (r, g, b):cs) = turtlePathLines (Turtle tp ta (Color3 (r / 255) (g / 255) (b / 255))) cs
        turtlePathLines t (StartBranch:cs) = (inBranchLines ++ afterThisBranchLines, maybeAfterThisBranchCs)
            where
                (inBranchLines, maybeAfterBranchCs) = turtlePathLines t cs
                (afterThisBranchLines, maybeAfterThisBranchCs) = turtlePathLines t (fromJust maybeAfterBranchCs)
        turtlePathLines t (EndBranch:cs) = ([], Just cs)

scaleLine :: Float -> Line -> Line
scaleLine k (Line c v1 v2) = Line c (scaleV k v1) (scaleV k v2)
    where
        scaleV k (Vertex2 vx vy) = Vertex2 (vx * k) (vy * k)

rescaleLines :: [Line] -> [Line]
rescaleLines lines = map (scaleLine k) lines
    where
        maxD = foldr (\(Line c (Vertex2 v1x v1y) (Vertex2 v2x v2y)) p -> maximum $ map abs [p, v1x, v1y, v2x, v2y]) 0 lines
        k = 0.9 * (1 / maxD)

renderLine :: Line -> IO ()
renderLine (Line col ver1 ver2)= renderPrimitive Lines $ do
    color col
    vertex ver1
    vertex ver2

showCommands :: [Command] -> IO ()
showCommands commands = do
    initialWindowSize $= Size 800 800
    (_progName, _args) <- getArgsAndInitialize
    _window <- createWindow "Showing a lsit of commands"
    displayCallback $= showCommandsDisplay commands
    actionOnWindowClose $= ContinueExecution
    mainLoop

showCommandsDisplay :: [Command] -> DisplayCallback
showCommandsDisplay commands = do
    clear [ColorBuffer]
    mapM_ renderLine $ rescaleLines $ commandsToLines commands
    flush

showLSystem :: Ord a => Int -> LSystem a Command -> IO ()
showLSystem startN lsys = do
    initialWindowSize $= Size 800 800
    (_progName, _args) <- getArgsAndInitialize
    _window <- createWindow "Showing a LSystem"

    nIORef <- newIORef (0 :: Int)
    nIORef $= startN

    displayCallback $= showLSystemDisplay lsys nIORef
    keyboardCallback $= Just (showLSystemKeyboard nIORef)
    mouseWheelCallback $= Just (showLSystemMouseWheel nIORef)
    actionOnWindowClose $= ContinueExecution

    mainLoop

showLSystemKeyboard :: IORef Int -> KeyboardCallback
showLSystemKeyboard nIORef c _ = case c of
    '-' -> nIORef $~ max 0 . (+(-1)) >> postRedisplay Nothing
    '+' -> nIORef $~ (+1) >> postRedisplay Nothing
    _ -> return ()

showLSystemMouseWheel :: IORef Int -> MouseWheelCallback
showLSystemMouseWheel nIORef _ d _ = nIORef $~ max 1 . (+d) >> postRedisplay Nothing

showLSystemDisplay :: Ord a => LSystem a Command -> IORef Int -> DisplayCallback
showLSystemDisplay lsys nIORef = do
    clear [ColorBuffer]
    n <- get nIORef
    let lines = commandsToLines $ translateAfterNSteps lsys n
    mapM_ renderLine $ rescaleLines lines
    -- print n white, bottom left corner
    color $ (Color3 1 1 1 :: Color3 Float)
    currentRasterPosition $= Vertex4 (-1) (-1) 0 1
    renderString TimesRoman24 (show n)
    currentRasterPosition $= Vertex4 (0.5) (-1) 0 1
    renderString TimesRoman24 (show (length lines))
    flush


app/Main.hs
module Main where

import System.Random
import Control.Monad
import System.Console.ArgParser

import LSystem
import LSystemRenderer

import Graphics.UI.GLUT

data Action = RunRandomLSystem | ShowLSystem Bool Bool Int Float deriving (Eq, Show)

actionParser :: IO (CmdLnInterface Action)
actionParser = mkSubParser [
    ("random", setAppDescr (mkDefaultApp (pure RunRandomLSystem) "random")
        "Keep generating and showing random LSystems"),
    ("show", setAppDescr (mkDefaultApp (
                                        ShowLSystem `parsedBy` 
                                        boolFlag "record" `Descr` "Get input in record syntax (the same way it is printed out)." `andBy`
                                        boolFlag "normal" `Descr` "Assume F means go forward, + turn right, - turn left and [] indicates branching" `andBy`
                                        optFlag 5 "steps" `Descr` "The number of steps at which should the program begin." `andBy`
                                        optFlag 90 "angle" `Descr` "The angle to use when the `standard` option is used"
                                       ) "show")
        "Show a specific LSystem. It will be read of standard input")
                             ]

doAction :: Action -> IO ()
doAction RunRandomLSystem = generateAndShowLSystemsForever
doAction (ShowLSystem record standard startN angle) = showLSystemFromStdInput record standard startN angle

main = do
    interface <- actionParser
    runApp interface doAction

--------------------------------------------------------------------------------
-- RunRandomLSystem
--------------------------------------------------------------------------------
-- This is the old amazing magick
-- randomNVariation n set = replicateM n (((!!) <$> pure set) <*> randomRIO (0, length set - 1))
randomNVariation :: Int -> [a] -> IO [a]
randomNVariation n set = replicateM n ((set !!) <$> randomRIO (0, length set - 1))

randomVariationBounded :: Int -> Int -> [a] -> IO [a]
randomVariationBounded lo hi set = do
    n <- randomRIO (lo, hi)
    randomNVariation n set

randomVarRulesBounded :: Int -> Int -> [a] -> [b] -> IO (Rules a b)
randomVarRulesBounded lo hi vars vcs = mapM (\x -> liftMonadFromTuple (return x, randomVariationBounded lo hi vcs)) vars
    where
        liftMonadFromTuple = uncurry $ liftM2 (,)

randomCommand :: IO Command
randomCommand = do
    x <- randomRIO (0, 2) :: IO Int
    case x of
        0 -> do
            d <- randomRIO (1, 2) :: IO Int
            return (Go (10 * fromIntegral d))
        1 -> do
            a <- randomRIO (1, 23) :: IO Int
            return (Turn (15 * fromIntegral a))
        2 -> do
            r <- randomRIO (0, 255) :: IO Int
            g <- randomRIO (0, 255) :: IO Int
            b <- randomRIO (0, 255) :: IO Int
            return (SetColor (fromIntegral r, fromIntegral g, fromIntegral b))

randomCommandsOfLen :: Int -> IO [Command]
randomCommandsOfLen 0 = return []
randomCommandsOfLen n = do
    maybeBranch <- randomRIO (0, 10) :: IO Int
    case maybeBranch of
        0 -> do
            branchLength <- randomRIO (1, n)
            branchCommands <- replicateM branchLength randomCommand
            afterBranchCommands <- randomCommandsOfLen (n - branchLength)
            return $ [StartBranch] ++ branchCommands ++ [EndBranch] ++ afterBranchCommands
        _ -> do
            newRandomCommand <- randomCommand
            remainingCommands <- randomCommandsOfLen (n - 1)
            return $ newRandomCommand : remainingCommands

getRandomLSystem :: Int -> Int -> Int -> Int -> Int -> Int -> IO (LSystem Char Command)
getRandomLSystem vCount cCount aCount ruleMin ruleMax transLen
  | vCount + cCount > 27 = error "There is not enough letters in the alphabet"
  | otherwise = do
    let vars = take vCount ['A'..'Z']
        cons = take cCount (filter (not . (`elem` vars)) ['A'..'Z'])
    axiom <- randomNVariation aCount vars
    rules <- randomVarRulesBounded ruleMin ruleMax vars (vars ++ cons)
    translationRules <- mapM (\x -> randomCommandsOfLen transLen >>= (\c -> return (x, c))) (vars ++ cons)
    return (LSystem vars cons axiom rules translationRules)

generateAndShowLSystemsForever :: IO ()
generateAndShowLSystemsForever = do
    lsys <- getRandomLSystem 2 2 1 5 15 5
    print lsys
    showLSystem 5 lsys
    generateAndShowLSystemsForever

--------------------------------------------------------------------------------
-- ShowLSystem
--------------------------------------------------------------------------------
showLSystemFromStdInput :: Bool -> Bool -> Int -> Float -> IO ()
showLSystemFromStdInput record standard startN angle =
    if record
        then do
            lsys <- readLn :: IO (LSystem Char Command)
            showLSystem startN lsys
        else do
            vars <- getLine
            cons <- getLine
            axiom <- getLine
            rules <- readLn :: IO [(Char, [Char])]
            transRules <- if standard
                             then return [('F', [Go 10]), ('+', [Turn angle]), ('-', [Turn (-angle)]), ('[', [StartBranch]), (']', [EndBranch])]
                             else readLn :: IO [(Char, [Command])]
            let lsys = LSystem vars cons axiom rules transRules
            print lsys
            showLSystem startN lsys

-- Testing

testing :: IO ()
testing = do
    (_progName, _args) <- getArgsAndInitialize
    _window <- createWindow "Hello World"
    displayCallback $= display
    actionOnWindowClose $= ContinueExecution
    mainLoop
 
display :: DisplayCallback
display = do
    clear [ ColorBuffer ]
    renderPrimitive Lines $ do
        color (Color3 (0.5) (0.5) (0.5) :: Color3 GLfloat)
        vertex (Vertex2 0.5 0.7 :: Vertex2 Float)
        vertex (Vertex2 (-0.5) (-0.7) :: Vertex2 Float)
    flush
